reps/2023-10-13-accelerator-support.md

```diff
@@ -0,0 +1,399 @@
+# Ray Accelerator Support
+
+## Summary
+
+Holistic design of supporting accelerators in Ray.
+This also serves as the reference doc for contributors
+to add new accelerator support in the future.
+
+### General Motivation
+
+Nowadays, more and more specialized accelerators (e.g. TPU, HPU) are coming out to speed up AI workloads.
+Supporting those accelerators natively (similar to how Ray supports Nvidia GPUs) in Ray will increase Ray's adoption as the computing framework for scaling AI applications.
+
+### Should this change be within `ray` or outside?
+
+Inside `ray` project since we want to support those accelerators out of the box.
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@pcmoritz, @ericl, @scv119
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@scv119
+
+## Design
+
+### Ray Core
+
+#### API
+Users will use accelerators through Ray resources. For GPU accelerators (e.g. Nvidia, AMD, Intel GPUs), they will use the `GPU` resource name. For other accelerators, they will each have their own resource name (e.g. TPU accelerator has resource name `TPU`).
+
+| Resource    | Accelerator               |
```

Comment:
Is the distinction between GPU, TPU, HPU, etc. important? Can't we use a more generic term like "num_accelerators"? It also seems a bit inconsistent to have "num_gpus", but not "num_tpus", etc. - I'm guessing it is because you don't want too many of those "num_*" attributes.

You do mention that "num_accelerators" was an alternative you considered - but discarded it. The alternative you have chosen to special case GPUs with num_gpus but not for the others seems like it could do with a bit more cleanup.

Reply:
The H100 datasheet says this:

Note that the H100 GPUs are primarily built for executing
datacenter and edge compute workloads for AI, HPC, and data analytics, but not graphics
processing.

So blurs the distinction between GPU and other AI processors.

Reply:
We try to have the resource name match how these accelerators are being called by people.

> Is the distinction between GPU, TPU, HPU, etc. important?

Different accelerators are quite different: e.g. some supports fractional resource, some not. Also the user's code normally only works for one accelerator family since different accelerator family provides different programming model (e.g. Nvidia GPU uses cuda).
This means if we do `num_accelerators`, we also need to introduce an `accelerator_family` node label so uses can express the equivalent of `num_gpus=1` with `num_accelerators=1, accelerator_family="GPU"`